packages/@webex/internal-plugin-metrics/src/call-diagnostic/call-diagnostic-metrics-latencies.ts

```diff
@@ -0,0 +1,290 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable valid-jsdoc */
+
+import {MetricEventNames} from '../metrics.types';
+
+// we only care about client event and feature event for now
+
+/**
+ * @description Helper class to store latencies timestamp and to calculate various latencies for CA.
+ * @exports
+ * @class CallDiagnosticLatencies
+ */
+export default class CallDiagnosticLatencies {
+  latencyTimestamps: Map<MetricEventNames, number>;
+
+  /**
+   * @constructor
+   */
+  constructor() {
+    this.latencyTimestamps = new Map();
+  }
+
+  /**
+   * Clear latencies
+   */
+  public clearLatencies() {
+    this.latencyTimestamps.clear();
+  }
+
+  /**
+   * Store Latency value
+   * @param key - key
+   * @param  value -value
+   * @throws
+   * @returns
+   */
+  public saveTimestamp(key: MetricEventNames, value: number = new Date().getTime()) {
+    this.latencyTimestamps.set(key, value);
+  }
+
+  /**
+   * Helper to calculate end - start
+   * @param a start
+   * @param b end
+   * @returns latency
+   */
+  public getDiffBetweenTimestamps(a: MetricEventNames, b: MetricEventNames) {
+    const start = this.latencyTimestamps.get(a);
+    const end = this.latencyTimestamps.get(b);
+    if (start && end) {
+      return end - start;
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Meeting Info Request
+   * @returns - latency
+   */
+  public getMeetingInfoReqResp() {
+    return this.getDiffBetweenTimestamps(
+      'client.meetinginfo.request',
+      'client.meetinginfo.response'
+    );
+  }
+
+  /**
+   * Interstitial Time
+   * @returns - latency
+   */
+  public getShowInterstitialTime() {
+    return this.getDiffBetweenTimestamps(
+      'client.interstitial-window.launched',
+      'client.meeting.click.joinbutton'
+    );
+  }
+
+  /**
+   * Call Init Join Request
+   * @returns - latency
+   */
+  public getCallInitJoinReq() {
+    return this.getDiffBetweenTimestamps(
+      'client.meeting.click.joinbutton',
+      'client.locus.join.request'
+    );
+  }
+
+  /**
+   * Locus Join Request
+   * @returns - latency
+   */
+  public getJoinReqResp() {
+    return this.getDiffBetweenTimestamps('client.locus.join.request', 'client.locus.join.response');
+  }
+
+  /**
+   * Locus Join Response Sent Received
+   * @returns - latency
+   */
+  public getJoinRespSentReceived() {
+    // TODO: not clear
```

Comment:
do we have a jira or something to make sure it's not forgotten?

Reply:
I have these on my notes, I haven't created actual JIRAs.

Reply:
[SPARK-440554](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-440554)